### ISSUE TITLE
Core Validators compare Strings

### ISSUE THREAD

**Issue by franciscocabral** [yiisoft/yii2#7365]
I have this rule:

```php
public function rules() {
  return [
    ['begin_date', \yii\validators\CompareValidator::className(), 
              'type' => 'string',
              'compareValue'=>(string)date("Y-m-d"),
              'operator' => '>',
              'message'=>'{value} must be greater than {compareValue}.'
    ],
  ];
}
```
I need to compare if my "begin_date" is greater than today, but I'm get some weird response. 
Like:
If my begin_date is "2015-02-20" and today is "2015-02-19", this says that my string isn't greater.
If my begin_date is "2015-03-01" and today is "2015-02-19", this says that my string isn't greater.
If my begin_date is "2016-01-01" and today is "2015-02-19", this says that my string is greater.

But if I put "==" instead of ">" in the operator, my begin_date need to be the exactaly same date as today.
Why with one operator it checks everything, but with other doesn't?

//sorry for my english

**Comment by lynicidn** [comment]
```php
['being_date', 'filter', 'filter' => 'strtotime'],
['being_date', 'compare', 'type' => 'number', 'compareValue' => time(), 'operator' => '>'],
```
http://en.code-bude.net/2013/12/17/how-to-compare-dates-in-php/

**Comment by lynicidn** [comment]
or use https://github.com/yiisoft/yii2/blob/master/framework/validators/DateValidator.php#L61

**Comment by franciscocabral** [comment]
@lynicidn the approach didn't work for me.
When I put "type"="number" I recive only the year.
The dateValidator can't be used to compare dates(can it?).

I'm using this in my form:
```php
     $form->field($model, 'begin')->input("date");
```

this are my rules:
```php
    public function rules() {
        return [
            [['coordinator'], 'integer'],
            [['name', 'begin', 'end'], 'required'],
            [['begin', 'end'], 'safe'],
            [['begin', 'end'], 'date'],
            [['name'], 'string', 'max' => 20]
        ];
    }
```

Would be greate if the compare dates was a dateValidator option, in fact, each validator should have your own compare option with suport to regex.

**Comment by lynicidn** [comment]
i don't show `number` in ur code

**Comment by franciscocabral** [comment]
when I put this:
```php
[['begin', 'end'], 'string'],
['being', 'filter', 'filter' => 'strtotime'],
['being', 'compare', 'type' => 'number', 'compareValue' => time(), 'operator' => '>'],
```
my begin gets only the year and not the strtotime of that date

**Comment by lynicidn** [comment]
@franciscocabral 
need `being` or `begin` ?

**Comment by franciscocabral** [comment]
sorry, worng typing here, didn't ctrl+c/v

**Comment by lynicidn** [comment]
show real ur code

**Comment by franciscocabral** [comment]
My rules:
```php

    public function rules() {
        return [
            [['coordinator'], 'integer'],
            [['name', 'begin', 'end'], 'required'],
            [['begin', 'end'], 'safe'],
            [['begin', 'end'], 'string'],
            ['begin', 'filter', 'filter' => 'strtotime'],
            ['begin', 'compare', 'type' => 'number', 'compareValue' => time(), 'operator' => '>',
                'message'=>'{value} must be greater than {compareValue}.'],
            [['name'], 'string', 'max' => 20]
        ];
    }
```

My form
```php
    <?= $form->field($model, 'coordinator')->hiddenInput() ?>
    <?= $form->field($model, 'name')->textInput(['maxlength' => 20]) ?>
    <?= $form->field($model, 'begin')->input("date") ?>
    <?= $form->field($model, 'end')->input("date") ?>
```

**Comment by lynicidn** [comment]
1. remove [['begin', 'end'], 'safe'],
2. show submitted data and $model->errors

**Comment by franciscocabral** [comment]
Without 'safe':
```php
    public function rules() {
        return [
            [['coordinator'], 'integer'],
            [['name', 'begin', 'end'], 'required'],
            //[['begin', 'end'], 'safe'],
            [['begin', 'end'], 'string'],
            ['begin', 'filter', 'filter' => 'strtotime'],
            ['begin', 'compare', 'type' => 'number', 'compareValue' => time(), 'operator' => '>',
                'message'=>'{value} must be greater than {compareValue}.'],
            [['name'], 'string', 'max' => 20]
        ];
    }
```

```
2015 must be greater than 1424440453.
```

**Comment by lynicidn** [comment]
show submitted data

**Comment by franciscocabral** [comment]
Can't submit, is cheking before.
But in the console I got this:
```Javascript
$("#campaign-begin").val()
"2015-01-01"
```

**Comment by lynicidn** [comment]
change:
```php
['begin', 'filter', 'filter' => 'strtotime'],
```
to
```php
['begin', 'date', 'format' => 'php:Y-m-d', 'timestampAttribute' => 'begin']
```

if it don't work - use forum

**Comment by franciscocabral** [comment]
Didn't work. Thanks for your time.

I really think it is a bug, using compareValidator, the date is fucked...

**Comment by lynicidn** [comment]
it's not bug, in your code strtotime return invalid value
@franciscocabral show error again with current code

**Comment by lynicidn** [comment]
test cover it https://github.com/yiisoft/yii2/blob/master/tests/unit/framework/validators/DateValidatorTest.php#L122
and timestamp generate php function https://github.com/yiisoft/yii2/blob/master/framework/validators/DateValidator.php#L157

**Comment by franciscocabral** [comment]
Rules:
```php
    public function rules() {
        return [
            [['coordinator'], 'integer'],
            [['name', 'begin', 'end'], 'required'],
            //[['begin', 'end'], 'safe'],
            //[['begin', 'end'], 'string'],
            ['begin', 'date', 'format' => 'php:Y-m-d', 'timestampAttribute' => 'begin'],
            ['begin', 'compare', 'type' => 'number', 'compareValue' => time(), 'operator' => '>',
                'message'=>'{value} must be greater than {compareValue}.'],
            [['name'], 'string', 'max' => 20]
        ];
    }
```

Error:
```
2015 must be greater than 1424442509.
```

**Comment by franciscocabral** [comment]
Another test:
```
    public function rules() {
        return [
            [['coordinator'], 'integer'],
            [['name', 'begin', 'end'], 'required'],
            //[['begin', 'end'], 'safe'],
            [['begin', 'end'], 'string'],
            //['begin', 'date', 'format' => 'php:Y-m-d', 'timestampAttribute' => 'begin'],
            ['begin', 'filter', 'filter' => function($v){
                return $v;
            
            }],
            ['begin', 'compare', 'type' => 'number', 'compareValue' => time(), 'operator' => '>',
                'message'=>'{value} must be greater than {compareValue}.'],
            [['name'], 'string', 'max' => 20]
        ];
    }
 ```

Error:
```
2015 must be greater than 1424442705.
```

**Comment by lynicidn** [comment]
see network console for monitoring request (post data)
ctrl+shift+k in firefox

**Comment by lynicidn** [comment]
ur form send error value, not "2015-01-01"

or load not correct populate it

**Comment by lynicidn** [comment]
show $model->begin before validate

**Comment by franciscocabral** [comment]
```php
    public function beforeValidate() {
        var_dump($this->begin);exit;
        parent::beforeValidate();
    }
```
It doesn't execute...
the CTRL+SHIFT+K(Mozilla) or F12(Chrome) doesn't show any request.
the form isn't sent because the validation.

**Comment by lynicidn** [comment]
u have problem in ur code. Sorry i don't know solve for unknown error

**Comment by franciscocabral** [comment]
I found, was without Preserve Log...

header:
```
Campaign[coordinator]:1
Campaign[name]:teste
Campaign[begin]:2015-01-01
Campaign[end]:2015-01-01
```

**Comment by lynicidn** [comment]
u have not correct code https://github.com/yiisoft/yii2/issues/7365#issuecomment-75246382
i don't write it

**Comment by lynicidn** [comment]
```php
            //['begin', 'date', 'format' => 'php:Y-m-d', 'timestampAttribute' => 'begin'],
            ['begin', 'filter', 'filter' => function($v){
                return $v;

            }],
```

**Comment by cebe** [comment]
> The dateValidator can't be used to compare dates(can it?).

this feature has been added and will be released with 2.0.4:
https://github.com/yiisoft/yii2/commit/3f2495465940deab3d390538e46b17cfa4caf1a7#diff-32236b8409f422c4a80339d1ef6466e0R101

**Comment by franciscocabral** [comment]
thanks @cebe and @lynicidn, another moment I'll test it. =]
